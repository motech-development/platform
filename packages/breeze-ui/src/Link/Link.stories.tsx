import { withA11y } from '@storybook/addon-a11y';
import { text, withKnobs } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import React from 'react';
import { MemoryRouter } from 'react-router-dom';
import BaseStyles from '../BaseStyles/BaseStyles';
import Card from '../Card/Card';
import Link from './Link';

const stories = storiesOf('Link', module);

stories.addDecorator(withA11y);
stories.addDecorator(withKnobs);

stories.add('Basic link', () => (
  <MemoryRouter>
    <BaseStyles />

    <Card>
      <Link to={text('To', '/home')}>{text('Text', 'This is a link')}</Link>
    </Card>
  </MemoryRouter>
));
