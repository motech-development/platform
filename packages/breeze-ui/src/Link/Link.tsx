import { FC, memo } from 'react';
import { Link as RouterLink, LinkProps } from 'react-router-dom';
import BaseLink from '../BaseLink/BaseLink';

const Link: FC<LinkProps> = (props) => (
  // eslint-disable-next-line react/jsx-props-no-spreading
  <BaseLink as={RouterLink} {...props} />
);

export default memo(Link);
