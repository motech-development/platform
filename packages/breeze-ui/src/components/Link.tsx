import { ComponentPropsWithoutRef, createElement, ElementType } from 'react';
import { classNames } from '../utils/className';

interface ILinkProps<C extends ElementType> {
  as?: C;
}

export type TLinkProps<C extends ElementType> = ILinkProps<C> &
  Omit<ComponentPropsWithoutRef<C>, keyof ILinkProps<C>>;

const Link = <C extends ElementType = 'a'>({
  as,
  children,
  className = '',
  ...rest
}: TLinkProps<C>) => {
  const component = as || 'a';

  return createElement(
    component,
    {
      className: classNames(
        'appearance-none bg-none border-0 font-semibold text-blue-600 underline hover:no-underline hover:text-red-600 focus:outline-none',
        className,
      ),
      ...rest,
    },
    children,
  );
};

export default Link;
