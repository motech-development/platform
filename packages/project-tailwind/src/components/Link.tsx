import { ElementType } from 'react';
import { Box, PolymorphicComponentProps } from 'react-polymorphic-box';
import {
  TTheme,
  TSizing,
  Themes,
  Sizing,
  useTailwind,
} from '../utilities/tailwind';

/** Default component element */
const DEFAULT_ELEMENT = 'a';

/** Link component props */
export type TLinkProps<E extends ElementType> = PolymorphicComponentProps<
  E,
  {
    /** Renders component as a block element */
    block?: boolean;

    /** Component theme */
    theme?: TTheme;

    /** Component size */
    size?: TSizing;
  }
>;

/**
 * Flexible anchor element
 *
 * @param props - Link component props
 *
 * @returns Link component
 */
export function Link<E extends ElementType = typeof DEFAULT_ELEMENT>({
  block = false,
  className,
  theme = Themes.PRIMARY,
  size = Sizing.MD,
  ...rest
}: TLinkProps<E>) {
  const { createStyles } = useTailwind(theme, size);

  const linkStyles = createStyles({
    classNames: [
      'appearance-none bg-none border-0 font-semibold underline hover:no-underline focus:outline-none focus:ring-2 focus:ring-offset-2',
      {
        'block w-full text-center': block,
        'inline-block': !block,
      },
      className,
    ],
    sizing: {
      lg: ['text-lg'],
      md: ['text-base'],
      none: ['text-base'],
      sm: ['text-sm'],
    },
    theme: {
      danger: ['text-red-600 hover:text-red-700 focus:ring-red-500'],
      primary: ['text-blue-600 hover:text-blue-700 focus:ring-blue-500'],
      secondary: ['text-gray-600 hover:text-gray-700 focus:ring-gray-500'],
      success: ['text-green-600 hover:text-green-700 focus:ring-green-500'],
      warning: ['text-yellow-600 hover:text-yellow-700 focus:ring-yellow-500'],
    },
  });

  return <Box className={linkStyles} as={DEFAULT_ELEMENT} {...rest} />;
}
