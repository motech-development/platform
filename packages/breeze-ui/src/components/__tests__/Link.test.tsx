import { render } from '@testing-library/react';
import Link from '../Link';

describe('Link', () => {
  it('should correctly render the component', () => {
    const { container } = render(<Link href="/home">Link</Link>);

    expect(container.firstChild).toMatchSnapshot();
  });
});
