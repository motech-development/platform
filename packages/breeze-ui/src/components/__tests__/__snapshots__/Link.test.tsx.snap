// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Link should correctly render the component 1`] = `
<a
  class="appearance-none bg-none border-0 font-semibold text-blue-600 underline hover:no-underline hover:text-red-600 focus:outline-none"
  href="/home"
>
  Link
</a>
`;
